rusoto/services/s3/src/custom/util.rs

```diff
@@ -18,3 +23,145 @@ use rusoto_core::signature;
 pub fn encode_key<T: AsRef<str>>(key: T) -> String {
     signature::encode_uri_path(key.as_ref())
 }
+
+macro_rules! add_headers {
+    (
+        $input:ident , $req:ident ; $p:ident , $e:expr ; $( $t:tt )*
+    ) => (
+        {
+            if let Some(ref $p) = $input.$p {
+                $req.add_header($e, &$p.to_string());
+            }
+            add_headers! { $input, $req; $( $t )* }
+        }
+    );
+    (
+        $input:pat , $req:expr ;
+    ) => ({
+    });
+}
+
+
+macro_rules! add_params {
+    (
+        $input:ident , $params:ident ; $p:ident , $e:expr ; $( $t:tt )*
+    ) => (
+        {
+            if let Some(ref $p) = $input.$p {
+                $params.put($e, &$p);
+            }
+            add_params! { $input, $params; $( $t )* }
+        }
+    );
+    (
+        $input:pat , $req:expr ;
+    ) => ({
+    });
+}
+
+pub trait PreSignedRequest {
+    /// http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+    fn get_signed_url(&self, region: &Region, credentials: &AwsCredentials) -> String;
+}
+
+impl PreSignedRequest for GetObjectRequest {
+    /// https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html
+    fn get_signed_url(&self, region: &Region, credentials: &AwsCredentials) -> String {
+        let request_uri = format!("/{bucket}/{key}", bucket = self.bucket, key = self.key);
+        let mut request = SignedRequest::new("GET", "s3", &region, &request_uri);
+        let mut params = Params::new();
+
+        add_headers!(
```

Comment:
> I'm not 100% familiar with pre-signed urls, but is there a difference to "normal" requests in terms of which headers/params need to be included? Do you think there would be a way to re-use some of the existing/generated code, e.g. by exposing a pub(crate) method like as_signed_request on GetObjectRequest, etc?

Should I generate a PreSignedRequest automatically? like https://github.com/rusoto/rusoto/blob/master/service_crategen/src/commands/generate/codegen/rest_request_generator.rs#L8

Reply:
I was thinking this through earlier, and noticed that there might be subtle differences here between pre-signed urls and request signing when it comes to request payloads (such as for `PutObjectRequest`). In light of this, I think this is fine to have it manually for now.